### ISSUE TITLE
RUCSS: Automatic detection for dynamic DOM elements and styles

### ISSUE THREAD

**Issue by engahmeds3ed** [wp-media/wp-rocket#5129]
**Is your feature request related to a problem? Please describe.**
We need a way to automatically detect the dynamic DOM elements and styles to be preserved in the page.

**Describe the solution you'd like**
When we talked in the last meeting we found possible solutions which need to be validated based on the current issues/patterns as follows:-

1. Preserve alpha-numeric IDs/Classes which has more than number of characters (Here we will detect this pattern inside all internal CSS selectors of the page once we get the used CSS).
2. @CrochetFeve0251 's solution is to detect the hash when we clear the cache (the old hash and the new one) then replace the old one inside the usedCSS with the new one.
3. Prepare a list of the current preserved items and try to get a common pattern between all of them.

**Additional context**
Current preserved list:-
```
'.wp-container-',
'.wp-elements-',
'#wpv-expandable-',
'#ultib3-',
'.uvc-wrap-',
'.jet-listing-dynamic-post-',
'.vcex_',
'.wprm-advanced-list-',
```
and from third party compatibility:-
1. `.u7cb7434f55ddbc61829903e3a7b5fb85` https://github.com/wp-media/wp-rocket/issues/5041
2. Flatsome
```
'#section_',
'#text-box-',
'#banner-',
'#slider-',
'#gap-',
'#image_',
'#row-',
'#text-',
```

I'll try getting the full  exclusion list not the part that we detect to get the full idea.

**Comment by piotrbak** [comment]
Here's the list of the patterns:
https://docs.google.com/spreadsheets/d/1tIxvwSn_qJVdWIopqwVJ2DetAo5fNfIfr-d4ib1xs6k/edit?usp=sharing

**Comment by engahmeds3ed** [comment]
I got the main page builders examples and I found that the pattern of alpha-numeric is also there mainly without having dynamic part so if we safelisted all this pattern, we will have a big used CSS that isn't used in the page so I see that the cons is more than the pros.

@mostafa-hisham suggested to have the html from the SaaS and then the plugin serve it in the cache file but this will be valid for the first time we got the used CSS only and the next time the cache clears the HTML will change again and here we don't call the SaaS again so I believe this is partially valid solution.

I added some of the elements into the spreadsheet for elementor and visual composer.

I can't find any other solution for that from my search and thinking.

**Comment by remyperona** [comment]
Remember that the goal is not to have them in the used CSS, but to safelist the patterns when we remove the styles from the HTML.

**Comment by engahmeds3ed** [comment]
> Remember that the goal is not to have them in the used CSS, but to safelist the patterns when we remove the styles from the HTML.

Good point, so I need to check if those alpha-numeric patterns that are used extensively in elementor and visual composer as examples, are used in internal CSS or not, if not so I believe we could start with adding this alpha-numeric pattern to preserve the internal style into the page then we can test with main themes and popular page builders and see if the used CSS changes in a bad way or not.

checking that now.

**Comment by DahmaniAdame** [comment]
I've never seen any dynamic classes on an actual CSS file. So, we are mainly targeting inline CSS with dynamic selectors/classes. 

SaaS can be involved as we have already partially processed some information we need in this case. We store there an array of selectors when processing pages. 

We could use that to pick matching alpha-numeric patterns used on `classes` and `IDs`, strip the alpha-numeric part of the selector, and send them comma-separated as part of the payload to WP Rocket to be stored on the RUCSS table, or possibly on the filesystem as well, `rucss-exclusions.txt` for example, to avoid any impact on the database size. 

We then use them to be excluded from RUCSS using `rocket_rucss_inline_content_exclusions`.

This will result in less overhead to the website's server as we won't be doing any extensive `preg_match`ing to find the patterns on HTML, which will surely have a performance impact while processing pages. 

We can still remove false positives using that same filter `rocket_rucss_inline_content_exclusions`, and feed SaaS to dismiss the false positives automatically at its level.

**Comment by engahmeds3ed** [comment]
Thanks @DahmaniAdame 
> I've never seen any dynamic classes on an actual CSS file. So, we are mainly targeting inline CSS with dynamic selectors/classes.

Yes all of them are in style tags only.

> We could use that to pick matching alpha-numeric patterns used on classes and IDs, strip the alpha-numeric part of the selector, and send them comma-separated as part of the payload to WP Rocket to be stored on the RUCSS table, or possibly on the filesystem as well, rucss-exclusions.txt for example, to avoid any impact on the database size.

You mean if we have for example `banner-ou3hlkl45s` we will return `banner-` from the SaaS, right?
I don't know if this will affect the used CSS generation time or not and also will need to check this exclusions list every time we get used CSS from the SaaS side, so what about making it page specific list so we may have this list into the DB for each url and use the mentioned filter to exclude those patterns for the requested url only.

as a conclusion we have two options:-
1. Save the exclusions list in one place (filesystem file for example) and each time we have data from SaaS we will compare and make this list unique then apply this list on the whole pages.
2. Save this list for each url into the DB and apply it on each url separately.

What do u think? @mostafa-hisham @DahmaniAdame 

> We can still remove false positives using that same filter rocket_rucss_inline_content_exclusions, and feed SaaS to dismiss the false positives automatically at its level.

I don't know how this will happen automatically but we can think of it.

**Comment by DahmaniAdame** [comment]
> You mean if we have for example banner-ou3hlkl45s we will return banner- from the SaaS, right?

Yes! 

> I don't know if this will affect the used CSS generation time

There will be a slight increase. 

My thinking, since we already have the selectors extracted + SaaS has more resources than customers servers, it will cost less time and efforts overall. 

> so what about making it page specific list so we may have this list into the DB for each url and use the mentioned filter to exclude those patterns for the requested url only

Yes, it should be done for each URL. 

> as a conclusion we have two options

The first option looks good to me - grouping exclusions and applying them to all pages. 

If one dynamic selector is used by a theme, it's likely to be used on multiple pages.

To avoid an extensive cost on the customer's server, we could just append new selectors to a RUCSS exclusions file on the filesystem, and periodically run a command to keep only unique entries? 

To avoid extensive cost on SaaS, maybe we could add a flag to URLs where we already found dynamic selectors to tell the SaaS to only look for dynamic selectors after a set number of Used CSS regeneration?

The flag can be managed on the Used CSS table and sent as part of the RUCSS request.

**Comment by engahmeds3ed** [comment]
> If one dynamic selector is used by a theme, it's likely to be used on multiple pages.

The cases that I saw was that those CSS styles are being added with blocks so with the banner block for example they add some internal styles for it directly after the block itself so if you don't use the block in any page, you will not see those stylings.

**Comment by DahmaniAdame** [comment]
@engahmeds3ed, yes, this is how it usually goes. 

Nothing will change whether we go with a global exclusion file or a url-based exclusion file as we are using the `rocket_rucss_inline_content_exclusions` filter. 

Only inline style matching one of the selectors will be preserved.

**Comment by engahmeds3ed** [comment]
Here is the summary of the call with @mostafa-hisham talking about this issue and how SaaS may be involved in that:-

- We talked about how to detect this pattern and the process we thought about was to detect the alphanumeric classes and IDs from the internal CSS and return this list to the plugin and be saved with each url into the DB so the next time the cache is cleared we will (try to) get the alphanumeric items from the internal CSS but now from the page itself and compare it with the saved list to get the changed one(s) and preserve its style.
**SaaS will not be able to detect the pattern by himself as he has only the current page and doesn't have the previous page state to compare.**

- We are trying here to get an automatic solution to detect and preserve the patterns that changes with each cache clear without the need to update the plugin like the current solution so if this automatic detection will affect the performance we may take into consideration what we did and will do in the [Dynamic exclusion issue](https://github.com/wp-media/wp-rocket/issues/3602) so we can add the list of third parties in the backend and take one of the following solutions:-

1. We will make API call with the list of active plugins and the active theme to get the whole list for all optimizations then cache it to be used with the corresponding filters.
2. We will make third party compatibility class like what we do now and send a request with the plugin name to the backend to get only optimizations for this plugin then cache it to be used.

the advantages of the second solution over the first one is that the first one needs to send the request again with every plugin being activated or deactivated but the second one will only deletes its API's cache with the plugin deactivation and will be available if the plugin is activated by default.
Also the cons of the second solution is that we will need to create a class for each plugin and this will require a plugin release.


Those are our final thoughts about this issue, any other idea/concern is highly appreciated.
